### PR TITLE
Resolve issues concerning upgrading scikit-learn to latest version

### LINKED ISSUE #3798: Prepare to use the latest version of scikit-learn
Currently, when upgrading to the latest version of scikit-learn, training of some models takes long and in particular cases hangs. Investigating and resolving the cause of the hang is crucial towards upgrading the dependency. 

Related Issues and PRs:
- #3770 
- #3671 
- #3559 
- #3308

### PR DESCRIPTION
**Pull request by jpangas**
**Issue:**
Closes #3798 

**Comments**:
This PR works towards investigating and resolving the issues concerning the upgrade to sci-kit learn's latest version, which is currently 1.4.1

To help with investigation, each new push will trigger the training a regression model on TC.

Train on Taskcluster: regression

### CODE REVIEW

**Comment by marco-c** [comment]
As the task normally takes around 30 minutes, could you reduce the timeout to a shorter time? Maybe something like 45 minutes?

**Comment by suhaibmujahid** [comment]
@jpangas could you please try to disable cross-validation to test if that is the problem?

**Comment by jpangas** [comment]
@suhaibmujahid , On it right away.. Waiting for the results.

**Comment by jpangas** [comment]
It went one step further but seems to hang too. https://community-tc.services.mozilla.com/tasks/DMRsF4k5T_yDC_jXE1X4Mg

**Comment by marco-c** [comment]
It hanged on https://github.com/mozilla/bugbug/blob/0586c6a49732f03781daf43d8726b53e2049e3db/bugbug/model.py#L397, it didn't get to https://github.com/mozilla/bugbug/blob/0586c6a49732f03781daf43d8726b53e2049e3db/bugbug/model.py#L405.

Could you try removing the sampler at https://github.com/mozilla/bugbug/blob/0586c6a49732f03781daf43d8726b53e2049e3db/bugbug/models/defect.py#L27?

**Comment by marco-c** [comment]
@jpangas don't push merge commits, it will just retrain needlessly.

**Comment by jpangas** [comment]
Sorry about that! I had synced on Github, pulled locally and was working on adding the temporary `verbosity` param, as well as removing `SMOTE`

File: bugbug/model.py [hunk 1]
```diff
@@ -147,7 +147,7 @@ def __init__(self, lemmatization=False):
         else:
             self.text_vectorizer = TfidfVectorizer
 
-        self.cross_validation_enabled = True
```
Comment (suhaibmujahid) [inline bugbug/model.py:150]: Could you please enable back the cross-validation?

Comment (jpangas) [reply bugbug/model.py]: Interestingly, there is no hang this time.

**Comment by marco-c** [comment]
So it turns out the problem is the sampler. To debug further, we would need to add logging inside the imblearn package.

**Comment by jpangas** [comment]
There is something peculiar I discovered when trying to debug the pipeline. 
Take a look at the logs when I try to train `spambug` :

<details><summary>1. Using `BorderlineSMOTE` before the upgrade. (`scikit-learn==1.1.3`)</summary>

```
2023-11-13 18:51:33,279:INFO:bugbug.model:X: (50296, 23527), y: (50296,)
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Pipeline] ... (step 1 of 2) Processing borderlinesmote, total=  13.0s
[Pipeline] ..... (step 2 of 2) Processing xgbclassifier, total= 1.2min
[CV] END  accuracy: (test=0.994) precision: (test=0.981) recall: (test=0.952) total time= 1.4min
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:  1.4min remaining:    0.0s
[Pipeline] ... (step 1 of 2) Processing borderlinesmote, total=   6.3s
[Pipeline] ..... (step 2 of 2) Processing xgbclassifier, total= 1.2min
[CV] END  accuracy: (test=0.993) precision: (test=0.964) recall: (test=0.959) total time= 1.3min
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:  2.7min remaining:    0.0s
[Pipeline] ... (step 1 of 2) Processing borderlinesmote, total=   5.9s
[Pipeline] ..... (step 2 of 2) Processing xgbclassifier, total= 1.2min
[CV] END  accuracy: (test=0.992) precision: (test=0.971) recall: (test=0.942) total time= 1.3min
[Pipeline] ... (step 1 of 2) Processing borderlinesmote, total=   5.9s
[Pipeline] ..... (step 2 of 2) Processing xgbclassifier, total= 1.2min
[CV] END  accuracy: (test=0.992) precision: (test=0.967) recall: (test=0.951) total time= 1.3min
[Pipeline] ... (step 1 of 2) Processing borderlinesmote, total=   6.0s
[Pipeline] ..... (step 2 of 2) Processing xgbclassifier, total= 1.2min
[CV] END  accuracy: (test=0.993) precision: (test=0.978) recall: (test=0.946) total time= 1.3min
[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:  6.7min finished
2023-11-13 18:58:14,509:INFO:bugbug.model:Cross Validation scores:
2023-11-13 18:58:14,509:INFO:bugbug.model:Accuracy: f0.9926655538619237 (+/- 0.0012548574986873567)
2023-11-13 18:58:14,510:INFO:bugbug.model:Precision: f0.9722933298803488 (+/- 0.013019383885616555)
2023-11-13 18:58:14,510:INFO:bugbug.model:Recall: f0.9501396681838834 (+/- 0.012000429255017857)
2023-11-13 18:58:14,510:INFO:bugbug.model:X_train: (45266, 23527), y_train: (45266,)
2023-11-13 18:58:24,211:INFO:bugbug.model:resampled X_train: (81908, 23527), y_train: (81908,)
2023-11-13 18:58:24,212:INFO:bugbug.model:X_test: (5030, 23527), y_test: (5030,)
2023-11-13 18:59:59,143:INFO:bugbug.model:Model trained

```

</details>

<details><summary>2. Using `BorderlineSMOTE` after the upgrade. (`scikit-learn==1.3.2`)</summary>

```
2023-11-13 19:03:22,754:INFO:bugbug.model:X: (50296, 23527), y: (50296,)
[Pipeline] ... (step 1 of 2) Processing borderlinesmote, total= 3.4min
[Pipeline] ..... (step 2 of 2) Processing xgbclassifier, total=  30.7s
[CV] END  accuracy: (test=0.993) precision: (test=0.982) recall: (test=0.947) total time= 3.9min
[Pipeline] ... (step 1 of 2) Processing borderlinesmote, total= 3.5min
[Pipeline] ..... (step 2 of 2) Processing xgbclassifier, total=  29.8s
[CV] END  accuracy: (test=0.992) precision: (test=0.967) recall: (test=0.949) total time= 4.0min
[Pipeline] ... (step 1 of 2) Processing borderlinesmote, total= 3.4min
[Pipeline] ..... (step 2 of 2) Processing xgbclassifier, total=  31.2s
[CV] END  accuracy: (test=0.992) precision: (test=0.973) recall: (test=0.944) total time= 3.9min
[Pipeline] ... (step 1 of 2) Processing borderlinesmote, total= 3.5min
[Pipeline] ..... (step 2 of 2) Processing xgbclassifier, total=  29.7s
[CV] END  accuracy: (test=0.992) precision: (test=0.967) recall: (test=0.949) total time= 4.0min
[Pipeline] ... (step 1 of 2) Processing borderlinesmote, total= 3.4min
[Pipeline] ..... (step 2 of 2) Processing xgbclassifier, total=  29.4s
[CV] END  accuracy: (test=0.993) precision: (test=0.982) recall: (test=0.943) total time= 3.9min
2023-11-13 19:23:03,862:INFO:bugbug.model:Cross Validation scores:
2023-11-13 19:23:03,862:INFO:bugbug.model:Accuracy: f0.9924888266288227 (+/- 0.001017650429623279)
2023-11-13 19:23:03,863:INFO:bugbug.model:Precision: f0.9740408238317727 (+/- 0.013533804667493323)
2023-11-13 19:23:03,863:INFO:bugbug.model:Recall: f0.9464292531583292 (+/- 0.004698820037821267)
2023-11-13 19:23:03,863:INFO:bugbug.model:X_train: (45266, 23527), y_train: (45266,)
2023-11-13 19:28:34,555:INFO:bugbug.model:resampled X_train: (81908, 23527), y_train: (81908,)
2023-11-13 19:28:34,555:INFO:bugbug.model:X_test: (5030, 23527), y_test: (5030,)
2023-11-13 19:29:09,020:INFO:bugbug.model:Model trained

```

</details>

<details><summary>3. Using `SMOTE` with the upgrade. (`scikit-learn==1.3.2`)</summary>

```
2023-11-13 18:40:23,702:INFO:bugbug.model:X: (50296, 23527), y: (50296,)
[Pipeline] ............. (step 1 of 2) Processing smote, total=  19.7s
[Pipeline] ..... (step 2 of 2) Processing xgbclassifier, total=  32.2s
[CV] END  accuracy: (test=0.994) precision: (test=0.983) recall: (test=0.950) total time=  51.9s
[Pipeline] ............. (step 1 of 2) Processing smote, total=  14.6s
[Pipeline] ..... (step 2 of 2) Processing xgbclassifier, total=  26.9s
[CV] END  accuracy: (test=0.993) precision: (test=0.967) recall: (test=0.956) total time=  41.5s
[Pipeline] ............. (step 1 of 2) Processing smote, total=  14.9s
[Pipeline] ..... (step 2 of 2) Processing xgbclassifier, total=  27.3s
[CV] END  accuracy: (test=0.993) precision: (test=0.978) recall: (test=0.948) total time=  42.3s
[Pipeline] ............. (step 1 of 2) Processing smote, total=  14.9s
[Pipeline] ..... (step 2 of 2) Processing xgbclassifier, total=  27.5s
[CV] END  accuracy: (test=0.992) precision: (test=0.973) recall: (test=0.947) total time=  42.4s
[Pipeline] ............. (step 1 of 2) Processing smote, total=  18.9s
[Pipeline] ..... (step 2 of 2) Processing xgbclassifier, total=  32.5s
[CV] END  accuracy: (test=0.993) precision: (test=0.980) recall: (test=0.944) total time=  51.5s
2023-11-13 18:44:13,345:INFO:bugbug.model:Cross Validation scores:
2023-11-13 18:44:13,346:INFO:bugbug.model:Accuracy: f0.9929306593518165 (+/- 0.0008841717953546909)
2023-11-13 18:44:13,346:INFO:bugbug.model:Precision: f0.9761892826703618 (+/- 0.011340875594534816)
2023-11-13 18:44:13,346:INFO:bugbug.model:Recall: f0.948980381929975 (+/- 0.007878938045211073)
2023-11-13 18:44:13,346:INFO:bugbug.model:X_train: (45266, 23527), y_train: (45266,)
2023-11-13 18:44:38,063:INFO:bugbug.model:resampled X_train: (81908, 23527), y_train: (81908,)
2023-11-13 18:44:38,063:INFO:bugbug.model:X_test: (5030, 23527), y_test: (5030,)
2023-11-13 18:45:10,597:INFO:bugbug.model:Model trained

```

</details>

One thing I was wondering is could this be an issue with `imbalanced-learn`? It takes about 20s with `SMOTE` while it takes about 3mins 50 s with `BorderlineSMOTE` using sklearn version `1.3.2`  from about 6s in version `1.1.3`. I dont think it is normal to have such a significant performance  difference when fitting with `BorderlineSMOTE` and `SMOTE`?

**Comment by marco-c** [comment]
It seems worth opening an issue in the imbalanced-learn repository at https://github.com/scikit-learn-contrib/imbalanced-learn, if we can build a small reproducible test case for them.

**Comment by jpangas** [comment]
I opened an issue with imbalanced-learn: https://github.com/scikit-learn-contrib/imbalanced-learn/issues/1056

**Comment by glemaitre** [comment]
Follow-up there: https://github.com/scikit-learn/scikit-learn/issues/28191

**Comment by jpangas** [comment]
Some tests are expected to fail because of deprecated methods and args. I will submit a fix for them.

**Comment by suhaibmujahid** [comment]
Thank you, @jpangas! It seems that `v1.4.1.post1` is addressing https://github.com/scikit-learn/scikit-learn/issues/28191: https://scikit-learn.org/stable/whats_new/v1.4.html#changes-impacting-many-modules

**Review by suhaibmujahid — APPROVED** [review]

**Pull request merged**